### PR TITLE
Adding Waspleau example project

### PR DESCRIPTION
**Pull request by shayneczyzewski**
# Description

Adds a new example to our examples list

- [ ] Code cleanup
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### CODE REVIEW

File: examples/waspleau/Notes.md [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+The fix was to add a dummy migration just to get past the Prisma error.
+
+
+| Bug captured here: https://github.com/wasp-lang/wasp/issues/398 |
+| --------------------------------------------------------------- |
+
+## 2) Axios version pinning
+
+```
+wasp: Error: Dependency conflict for user npm dependency (axios, ^0.21.1): Version must be set to the exactly the same version as the one wasp is using: ^0.21.1
```
Comment (Martinsos) [inline examples/waspleau/Notes.md:63]: How did you fix this one? From what I see you did use the exact same version as Wasp does, but Wasp still wasn't happy?

File: examples/waspleau/Notes.md [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+CallStack (from HasCallStack):
+  error, called at src/Generator/WebAppGenerator.hs:73:31 in waspc-0.2.2.2-B3XOTY4be7eFFZyFqJsQfH:Generator.WebAppGenerator
+```
+
+## 3) Dockerfile keeps getting overwritten. Any way to add to it? Do we want to use Docker Compose to allow for additional services like Redis?
```
Comment (Martinsos) [inline examples/waspleau/Notes.md:68]: Ha good question -> for now there is no way to add to it. Ideally we would have support for this in Wasp, if needed. What was your use case? You wanted to spin up a Redis as a separate micro service? Or just install it on the machine that server runs on?

File: examples/waspleau/Notes.md [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+
+## 4) Dockerfile python stale/heroku container does not support M1/aarch64 [`heroku container:push` did not work]
```
Comment (Martinsos) [inline examples/waspleau/Notes.md:70]: This is container that you provided yourself, by customizing the Dockerfile?

File: examples/waspleau/Notes.md [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+Snip to save: ` docker buildx build --platform linux/amd64 . -t web && docker tag web registry.heroku.com/waspleau/web && docker push registry.heroku.com/waspleau/web && heroku container:release --app waspleau web`
+
+## 5) When I pushed to Heroku I had a migration_lock.toml issue
+
+```
+2021-12-16T02:00:14.338320+00:00 app[web.1]: Prisma schema loaded from ../db/schema.prisma
+2021-12-16T02:00:14.366442+00:00 app[web.1]: Datasource "db": PostgreSQL database "dbne5egge1dhr8", schema "public" at "ec2-184-73-25-2.compute-1.amazonaws.com:5432"
+2021-12-16T02:00:14.564581+00:00 app[web.1]: Error: P3014
+2021-12-16T02:00:14.564591+00:00 app[web.1]:
+2021-12-16T02:00:14.564592+00:00 app[web.1]: The datasource provider `postgresql` specified in your schema does not match the one specified in the migration_lock.toml, `sqlite`. Please remove your current migration directory and start a new migration history with prisma migrate dev. Read more: https://pris.ly/d/migrate-provider-switch
```
Comment (Martinsos) [inline examples/waspleau/Notes.md:110]: Aha, this is because you switched from sqlite to postgresql -> yeah, once you do that you need to drop the existing migrations, that is by design (by Prisma). I guess we didn't mention this in the docs anywhere?

File: examples/waspleau/Notes.md [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+2021-12-16T02:18:54.254783+00:00 app[web.1]: npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
+2021-12-16T02:18:54.258329+00:00 app[web.1]:
+2021-12-16T02:18:54.258473+00:00 app[web.1]: npm ERR! A complete log of this run can be found in:
+2021-12-16T02:18:54.258545+00:00 app[web.1]: npm ERR!     /app/server/.npm/_logs/2021-12-16T02_18_54_255Z-debug.log
+2021-12-16T02:18:54.433588+00:00 heroku[web.1]: Process exited with status 1
+2021-12-16T02:18:54.568000+00:00 heroku[web.1]: State changed from starting to crashed
+2021-12-16T02:19:01.210559+00:00 app[api]: Deployed web (f5ce2d583e48) by user thelastemail@gmail.com
+```
+
+Had to delete and re-add database. Probably when I originally deployed with sqlite there it got into a funky state.
```
Comment (Martinsos) [inline examples/waspleau/Notes.md:178]: Oh wow hm!

File: examples/waspleau/Notes.md [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+
+```
+ 2021-12-16T02:34:30.070430+00:00 heroku[web.1]: Error R10 (Boot timeout) -> Web process failed to bind to $PORT within 60 seconds of launch
+2021-12-16T02:34:30.110821+00:00 heroku[web.1]: Stopping process with SIGKILL
+2021-12-16T02:34:30.279723+00:00 heroku[web.1]: Process exited with status 137
+2021-12-16T02:34:30.350174+00:00 heroku[web.1]: State changed from starting to crashed
+2021-12-16T02:34:31.160262+00:00 heroku[router]: at=error code=H10 desc="App crashed" method=GET path="/" host=waspleau.herokuapp.com request_id=ed3e548f-56a7-4042-8465-1fc684002d7f fwd="75.60.240.117" dyno= connect= service= status=503 bytes= protocol=https
+```
+
+Is it because the `.env` file is being copied into ./wasp/build/server? Or Redis? Or startup function?
```
Comment (Martinsos) [inline examples/waspleau/Notes.md:190]: `.env` file doesn't get copied anywhere, it is there only for development purposes.

So it turned out that REDIS_URL was missing -> ok that makes sense. How did you know that was the problem? The error message doesn't look very informative to me, I don't see it pointing to that in any way?

File: examples/waspleau/Notes.md [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+
+Needed to specify REDIS_URL on Heroku: `const queue = new Queue('waspleau', process.env.REDIS_URL || 'redis://127.0.0.1:6379')`
+
+However, do we want to be copying over the `.env` at all? Prob ok.
```
Comment (Martinsos) [inline examples/waspleau/Notes.md:194]: So right now, .env is supposed to be only for local development, and it is gitignored. You are supposed to manually ensure that ENV vars are correctly set on your hosting provider. That said, this is a space we might want to explore further, to possibly provide for a nicer experience.

File: examples/waspleau/Notes.md [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+
+## 8) App loads, but only renders the string "Hello world"
+
+Oh still need frontend deployed. Maybe instead of Hello world we note the frontend aspect somehow?
```
Comment (Martinsos) [inline examples/waspleau/Notes.md:198]: Ha good point! We could make it say "I am api-server" or smth like that, at least.

File: examples/waspleau/README.md [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+
+`DATABASE_URL="postgresql://postgres@localhost/waspleau-dev" wasp db migrate-dev`
+
+## Step 4
+
+`DATABASE_URL="postgresql://postgres@localhost/waspleau-dev" wasp start`
+
+This will start your background workers and present a dashboard UI that will auto-refresh every minute.
+
+Note: As you develop your own workers, keep in mind each time you save a file in the project it will automatically reload, hence restarting your server and server function. Use the `MOCK_DATA=true` command line envar if you want to avoid pinging external resources.
```
Comment (Martinsos) [inline examples/waspleau/README.md:25]: I guess what would help here is if we had hot reloading? I am not sure if it can be applied to this, but if it could, it would be cool.

File: examples/waspleau/ext/serverSetup.js [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+ * schedulign the repeated jobs using cron syntax.
+ * */
+import Queue from 'bull'
+
+const queue = new Queue('waspleau', process.env.REDIS_URL || 'redis://127.0.0.1:6379')
+
+const setupQueue = async () => {
+  // Clear entire queue when starting up.
+  // Warning: this is not cluster friendly nor a production best practice!
+  // This just helps keep our Redis queue tidy after lots of jobs/restarts.
```
Comment (Martinsos) [inline examples/waspleau/ext/serverSetup.js:71]: What is the good production practice then, how would we keep the queue tidy in that case?
Btw maybe you could check here for ENV that is being used, and then use that to avoid using this in production?

File: examples/waspleau/ext/serverSetup.js [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+}
+
+
+/**
+ * Dashboard Data
+ * 
+ * This is our in-memory data store we send down to the client via a query in dashboard.js.
+ * This of course means every time the server is restarted we lose this data and must repopulate.
+ * */
+const dashboardData = {}
```
Comment (Martinsos) [inline examples/waspleau/ext/serverSetup.js:21]: Coming from the functional programming world now, I feel uneasy seeing global state :D.

Idea: what if we put `dashboardData` and `workers` and `workerFunctions` in the anonymous function that you are exporting as default export? So they would be created inside it, and then it would return them as its return value? Do you think that would overly complicate things, or it could be an improvement?

**Review by Martinsos — APPROVED** [review]
@shayneczyzewski this is awesome, love the work you did with the workers and it is also great that we are hitting these new problems with this app! It seems like there is a lot we can learn from this, especially figuring out how to enable easy specifying of external services that need to run next to the backend, be it on the same api server or as a separate micro service.

I left some minor comments, and some questions.

Btw we should create issues for all of the challenges you encountered -> I believe you already created for some but I don't think you covered all of them?

File: examples/waspleau/Notes.md (hunk 1 above)
Comment (shayneczyzewski) [reply examples/waspleau/Notes.md:63]: This error was before that and it eventually worked. I can update the note to that effect. This was if I added the axios dependency without a version, it complained, so I had to pin to what Wasp used and then it went away.

File: examples/waspleau/Notes.md (hunk 2 above)
Comment (shayneczyzewski) [reply examples/waspleau/Notes.md:68]: I think there were two things to consider here, one being using docker compose for additional services like Redis or DB or nginx in front. The other is if for some reason the Docker file out of the box isn't quite working for the user, or they wanted to tweak it in some way. Not a must have, just something I noticed and to consider at some point in future.

File: examples/waspleau/Notes.md (hunk 3 above)
Comment (shayneczyzewski) [reply examples/waspleau/Notes.md:70]: Correct. We may have an actual issue in the generated Dockerfile, where `python` in `RUN apk add --no-cache build-base libtool autoconf automake python` no longer works and we need it to be `python3`. However, I am not 100% sure if this is M1 related or general. Could be a good test for someone else to verify.

Either way, I needed to update the Dockerfile and then use docker commands instead of heroku (but that part is specifically because of M1).

File: examples/waspleau/Notes.md (hunk 4 above)
Comment (shayneczyzewski) [reply examples/waspleau/Notes.md:110]: We do have it there: https://wasp-lang.dev/docs/language/basic-elements#migrating-from-sqlite-to-postgresql I will likely remove this part from my notes.

File: examples/waspleau/Notes.md (hunk 5 above)
Comment (shayneczyzewski) [reply examples/waspleau/Notes.md:178]: Yeah that part was super strange 🤔

File: examples/waspleau/Notes.md (hunk 6 above)
Comment (shayneczyzewski) [reply examples/waspleau/Notes.md:190]: Yeah in this case it was REDIS_URL, I will tidy that comment up. However, when I run `wasp build` with a .env file in my project, I do see it getting copied to .wasp/build/server:
Screen Shot 2021-12-21 at 8 26 42 AM

File: examples/waspleau/Notes.md (hunk 8 above)
Comment (shayneczyzewski) [reply examples/waspleau/Notes.md:198]: Yeah I think something like that may help!

File: examples/waspleau/README.md (hunk 9 above)
Comment (shayneczyzewski) [reply examples/waspleau/README.md:25]: Yeah I haven't looked into it much tbh but it would be nice to toggle in this use case.

File: examples/waspleau/ext/serverSetup.js (hunk 10 above)
Comment (shayneczyzewski) [reply examples/waspleau/ext/serverSetup.js:71]: Agree the comment here is not complete nor ideal. I was trying to convey in dev mode it is helpful (so we could wrap in an if), but it is also helpful and _probably_ ok in prod, but not really recommended by anyone. The two issues it solves are:
(1) addressed in comment, but removes old jobs. This can be done other ways though: https://github.com/OptimalBits/bull/issues/840 I am new to Bull and just didn't see that option before so I can use it and remove the comment!
(2) the bigger issue is that any time the recurring job name or cron schedule change, the old job is _not_ removed. So they could have multiple jobs running for the same logical one. So I could just have a note in there about that behavior and say whenever you do change either a job name or cron schedule to run this for safety (protected by an ENV VAR).

There are ways you can address (2) as well: https://github.com/OptimalBits/bull/blob/master/REFERENCE.md#queueremoverepeatable The hard part is I cannot really do that step for them since I won't know what the old job name was, only they will, I would have to remove them all anyways to provide actual safety.

File: examples/waspleau/ext/serverSetup.js (hunk 11 above)
Comment (shayneczyzewski) [reply examples/waspleau/ext/serverSetup.js:21]: Ah yes, I agree. It did make me uneasy as well. I can tinker with an approach like that. I also wonder from a JS GC perspective if I can get away with encapsulating by just putting all the variables and functions inside the anonymous function, since `queue.process('*'` has a callback that references `dashboardData` and `workerFunctions` they should not be garbage collected as they will be in the closure? I can test that part out, as well as see if returning them helps/differs any.

File: examples/waspleau/Notes.md (hunk 6 above)
Comment (shayneczyzewski) [reply examples/waspleau/Notes.md:190]: > How did you know that was the problem? The error message doesn't look very informative to me, I don't see it pointing to that in any way?

I had a feeling from the timeout it was stuck on an external service, but the Heroku logs don't provide a ton of detail. So dropped in some `console.log`s to confirm the Redis suspicion.

File: examples/waspleau/Notes.md (hunk 7 above)
Comment (shayneczyzewski) [reply examples/waspleau/Notes.md:194]: Yeah that makes sense. Although see above for my behavior that seems to show if you have a .env file in your project, when you run `wasp build` it does still get copied into .wasp/build/server/.env

**Comment by shayneczyzewski** [comment]
> @shayneczyzewski this is awesome, love the work you did with the workers and it is also great that we are hitting these new problems with this app! It seems like there is a lot we can learn from this, especially figuring out how to enable easy specifying of external services that need to run next to the backend, be it on the same api server or as a separate micro service.
> 
> I left some minor comments, and some questions.
> 
> Btw we should create issues for all of the challenges you encountered -> I believe you already created for some but I don't think you covered all of them?

Thanks for the review @Martinsos good feedback! I will clean up the notes, and then add any missing issues (although some are just nice to haves/for future reference). I will also clean up the `queue.obliterate` and see if I can not use the global vars. Appreciate it!

File: examples/waspleau/.gitignore [hunk 12]
```diff
@@ -0,0 +1,3 @@
+/.wasp/
+/.env
+.DS_Store
```
Comment (matijaSos) [inline examples/waspleau/.gitignore:3]: did you have to add this manually or was it already set here by Wasp? I ask b/c I don't remember adding it, and Martin is on Linux :D so we make sure it is added by default if it isn't atm

File: examples/waspleau/.wasproot [hunk 13]
```diff
@@ -0,0 +1 @@
+File marking the root of Wasp project.
```
Comment (matijaSos) [inline examples/waspleau/.wasproot:1]: We could also maybe add something like "Please never touch/delete this manually", or sth like that.

File: examples/waspleau/Notes.md [hunk 14]
```diff
@@ -0,0 +1,198 @@
...
+https://pris.ly/d/prisma-schema
+
+Error: Migrate dev failed: Migrate (dev) failed with exit code: 1
+```
+
+The fix was to add a dummy migration just to get past the Prisma error.
+
+
+| Bug captured here: https://github.com/wasp-lang/wasp/issues/398 |
+| --------------------------------------------------------------- |
```
Comment (matijaSos) [inline examples/waspleau/Notes.md:58]: nice formatting, love it! Didn't know github md supports this, really cool :).

File: examples/waspleau/Notes.md (hunk 4 above)
Comment (matijaSos) [reply examples/waspleau/Notes.md]: good info on missing that, probably will happen to others too. It might be cool in the future to have a mechanism that detects dev switched so could e.g. warn them through CLI. Def an over-optimisation for now, but it was just a thought :D.

File: examples/waspleau/Notes.md (hunk 5 above)
Comment (matijaSos) [reply examples/waspleau/Notes.md:167]: is this easy to repro? Would be good to create an issue if yes

File: examples/waspleau/ext/.waspignore [hunk 15]
```diff
@@ -0,0 +1,3 @@
+# Ignore editor tmp files
```
Comment (matijaSos) [inline examples/waspleau/ext/.waspignore:1]: we could also potentially scaffold for user a .waspignore file that covers the most common editors. Do you think it makes sense? If yes, we could make an issue and mark it as a "good-beginner-issue", we are always in lack of these :D.

File: examples/waspleau/ext/serverSetup.js [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+ * 
+ * Here we setup the Redis-backed Bull queue named 'waspleau'.
+ * Whenever we process a job in the queue's callback function, we look up the worker name and
+ * call the associated worker function. We then store the returned data to send to the client.
+ * 
+ * There is also a MOCK_DATA env var we check for and pass as an opt in case the worker
+ * wants to send mock data while doing active development.
+ * 
+ * Lastly, we populate the queue with a one-off job on startup for each worker, while also
+ * schedulign the repeated jobs using cron syntax.
```
Comment (matijaSos) [inline examples/waspleau/ext/serverSetup.js:62]:
```suggestion
 * scheduling the repeated jobs using cron syntax.
```

**Review by matijaSos — APPROVED** [review]
love it, this is a really cool example and I think we can also learn a lot from it :). Love how you documented your experience in detail, super helpful.

File: examples/waspleau/ext/serverSetup.js (hunk 11 above)
Comment (shayneczyzewski) [reply examples/waspleau/ext/serverSetup.js:21]: Ok, so I thought on it a bit more and am stumped on how to get rid of the top level declarations 😄  I think the hard part is the fact I need to access `dashboardData` in the queue callback, but then _also_ expose it to the query via some `export`ed function. I don't know that returning it from the default anonymous function would enable that second part (unless I missed part of the suggestion).

The good news is I think all modules are themselves wrapped in functions, so it is not global to all of node just this file. But if you have ideas for how we can do something like the following in a more locally scoped declaration happy to give it a go:

```javascript
// how can we declare dashboardData such that we use it in the queue callback
// registered in setupFn, but also export a way to get that data for a query?

export default async () => {
  queue.process('*', async (job) => {
    dashboardData[foo] = bar
  })
}

export const getDashboardData = () => Object.values(dashboardData).flat()
```

Thanks!

File: examples/waspleau/Notes.md (hunk 1 above)
Comment (shayneczyzewski) [reply examples/waspleau/Notes.md:65]: Updated here: https://github.com/wasp-lang/wasp/pull/412/commits/69fd53485f8f520c7c8c4066557e420271bfebd4

File: examples/waspleau/Notes.md (hunk 4 above)
Comment (shayneczyzewski) [reply examples/waspleau/Notes.md]: Others may read better than I do 🤓  lol but yes, I agree it would be great if in the future it detected the switch of database types and warned or did it for you 🚀

File: examples/waspleau/Notes.md (hunk 5 above)
Comment (shayneczyzewski) [reply examples/waspleau/Notes.md:167]: Yep, I can try to repro it sometime and will create an issue if so. I will add a TODO to remind me in case I cannot do before merge.

File: examples/waspleau/ext/serverSetup.js (hunk 10 above)
Comment (shayneczyzewski) [reply examples/waspleau/ext/serverSetup.js]: Updated here: https://github.com/wasp-lang/wasp/pull/412/commits/46cfc71a756ca2ac240d416d4ace85162b11f92a

File: examples/waspleau/.gitignore (hunk 12 above)
Comment (shayneczyzewski) [reply examples/waspleau/.gitignore:3]: When running `wasp new` we automatically get a `.gitignore` file with the following:
```
/.wasp/
/.env
```

Should I open an issue to add more things to the default `.gitignore` like `.DS_Store`?

File: examples/waspleau/.wasproot (hunk 13 above)
Comment (shayneczyzewski) [reply examples/waspleau/.wasproot:1]: Yeah I can create an issue for this update if you want!

File: examples/waspleau/ext/.waspignore (hunk 15 above)
Comment (shayneczyzewski) [reply examples/waspleau/ext/.waspignore:1]: I think that could be a good beginner issue. Would having editor ignores go into both the `.gitignore` and `.waspignore` perhaps?

File: examples/waspleau/Notes.md (hunk 2 above)
Comment (Martinsos) [reply examples/waspleau/Notes.md:78]: Makes sense, I think it is certainly something we have to consider in the future and figure out how to support it.

Should we create a separate issue for this? Or do we leave it here, assuming we will at some point revisit these notes and then create appropriate issues? I am not sure myself I have to admit.

File: examples/waspleau/Notes.md (hunk 3 above)
Comment (Martinsos) [reply examples/waspleau/Notes.md:80]: Uff ok, great job on succeeding! Same as above, not sure if we should create an issue now or leave it for later.

File: examples/waspleau/Notes.md (hunk 6 above)
Comment (shayneczyzewski) [reply examples/waspleau/Notes.md]: Yep will do!

Comment (shayneczyzewski) [reply examples/waspleau/Notes.md]: https://github.com/wasp-lang/wasp/issues/413

File: examples/waspleau/Notes.md (hunk 3 above)
Comment (shayneczyzewski) [reply examples/waspleau/Notes.md:80]: Thanks! How about I go ahead and create a [Waspleau example] issue and just have a list of some of these maybe things that we can either break out in the future into separate issues or ignore, mainly for tacking sake. Sound ok?

File: examples/waspleau/Notes.md (hunk 7 above)
Comment (Martinsos) [reply examples/waspleau/Notes.md]: No, I think that is a bug, commented above!

File: examples/waspleau/ext/serverSetup.js (hunk 10 above)
Comment (Martinsos) [reply examples/waspleau/ext/serverSetup.js]: Makes sense! It seems to me like it would be ok even for production, in our case, to clean the queue, since we don't really care much if an old job doesn't finish, as long as there is a new job coming after it, right? So clearing entire queue when starting up doesn't sound so bad to me. Ok, it depends on the type of workers I guess, but at the moment it is ok. Ok, no need to spend much time here / over-engineer, I just wanted to get some clarity on the topic, thanks!

File: examples/waspleau/ext/serverSetup.js (hunk 11 above)
Comment (Martinsos) [reply examples/waspleau/ext/serverSetup.js:21]: I know for sure that if you put the variable in the anonymous closure, it will live as long as it is being used, GC will not clean it up.

As for modules -> you are right, they are already encapsulated, so these variables are global only inside the module, and not through the whole codebase. Module is a like a big function that gets executed on import. I am not sure though if it can get executed multiple times, in case it is imported in multiple places -> I guess not, that would be problematic. So actually using these variables in this fashion as you did is not so bad, you are right, and probably it is not worth further optimizing and complicating. If I am correct, I believe we can look at JS module as a singleton.

Regarding using dashboardData in query -> you are right! They only way really is to keep some state in the module, and if we are keeping state, then it doesn't really matter if it is one variable or couple of them.

TLDR; You are right, modules are encapsualted / singletons and what you did is actually completely fine and probably the best way!

File: examples/waspleau/.gitignore (hunk 12 above)
Comment (Martinsos) [reply examples/waspleau/.gitignore:3]: .DS_Store is mac specific right? So maybe lets leave it out for now, and we will see, if it turns out people need it we can put it in.

File: examples/waspleau/ext/.waspignore (hunk 15 above)
Comment (Martinsos) [reply examples/waspleau/ext/.waspignore:1]: FYI @matijaSos , this .waspignore file you are looking at here is automatically generated when you create a new project, with these lines for ignoring editor tmp files.

**Review by Martinsos — APPROVED** [review]
@shayneczyzewski All is looking good, I approved!

I also commented on remaining conversations, I think that concludes most/all of them.

Regarding opening issues -> if you feel like it deserves an issue, do open one, but if not, leave it as it is and we will at some point revisit the notes here -> we know we have them because we do this for every example project, so we will not forget.

File: examples/waspleau/ext/serverSetup.js (hunk 10 above)
Comment (shayneczyzewski) [reply examples/waspleau/ext/serverSetup.js]: Yep no sweat, and I agree. Even in prod I think this should be fine, but yes there is the possibility of a race condition I think if you have it setup in a clustered setup where someone could start accepting while another boots and clears queue. I can make that final note in there, but agree, I didn't want to spend the time getting it to a 100% bulletproof solution to this corner case. Thanks

File: examples/waspleau/ext/serverSetup.js (hunk 11 above)
Comment (shayneczyzewski) [reply examples/waspleau/ext/serverSetup.js:21]: Cool thanks for the review, I appreciate it. I just did not want to miss an approach here, but it was an interesting use case to think about! Thanks

File: examples/waspleau/.gitignore (hunk 12 above)
Comment (shayneczyzewski) [reply examples/waspleau/.gitignore:3]: Yeah, that is Mac specific. Sounds good. I was going to open a beginner issue with a request for more files and patterns to add to `.gitignore` and `.waspignore` (both platform specific and perhaps editor based). Still worthwhile to do so to get some new contributors?

I was going to encourage them to look at: https://github.com/github/gitignore for inspiration

**Comment by shayneczyzewski** [comment]
> @shayneczyzewski All is looking good, I approved!
> 
> I also commented on remaining conversations, I think that concludes most/all of them.
> 
> Regarding opening issues -> if you feel like it deserves an issue, do open one, but if not, leave it as it is and we will at some point revisit the notes here -> we know we have them because we do this for every example project, so we will not forget.

Awesome, thanks @Martinsos ! I will plan to just create 1 tracking issue for all these _maybe_ ones we are discussing, mainly so I remember to revisit at some point in the future and either break them out into their own or ignore. Thanks again all

**Comment by shayneczyzewski** [comment]
I captured all the possible issues here in this combined issue for future tracking sake: https://github.com/wasp-lang/wasp/issues/415
